tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+                })
+
+            return return_mock_popen_obj
+        
+        action_scope_attributes = {}
+        for _, action in actionSet.items():
+            # The action passed to us is meant to return some sort of error_code.
+            # However, it is advantageous for our actions to return a mock specced to
+            # a popen object, so that our mock popen func can delegate straight to the
+            # action. see wrap_action_return_in_popen_obj for impl details.
+            action_returning_popen_obj = wrap_action_return_in_popen_obj(action["action"])
+            # Attempting to specify action["key"].side_effect throws an AttributeError
+            # at runtime, as we (below) have added a spec to the actionScope mock. Thus,
+            # we need to instantiate a mock object for the attribute directly.
+            action_scope_attributes[action["key"]] = Mock(wraps=action_returning_popen_obj)
+
+        actionScope.mock_add_spec(actionSet.keys()) # We rely on this spec to enable proper fallthrough (as described in patch_popen())
+        actionScope.configure_mock(**action_scope_attributes)
+    
+    @staticmethod
+    def _configure_data_fetcher(dataFetcher, actionSet, extensionInfo):
+        """
+        Replaces the "test_ext" fetcher function in the provided data fetcher
+        with a lambda which generates a proper zip file containing a HanderManifest.json
+        file populated with the correct extension metadata and command names (i.e. the
+        name of the commands to call for installCommand, enableCommand, etc.) from the
+        given actionSet and extensionInfo.
+        """
+
+        base_manifest = [{
+            "name": extensionInfo.name.split(".")[-1],
+            "version": extensionInfo.version,
+            "handlerManifest": {
+                "rebootAfterInstall": False, "reportHeartbeat": False,
+                "continueOnUpdateFailure": extensionInfo.continueOnUpdateFailure,
+                "updateMode": extensionInfo.updateMode
+            }
+        }]
+
+        for title, cmd in actionSet.items():
+            base_manifest[0]['handlerManifest'][title] = cmd['key']
+
+        dataFetcher["test_ext"] = mockwiredata.generate_ext_fetcher_func(base_manifest)
+
+
+    def __init__(self, dataFetcherBase, actionSet, extensionInfo):
+        """
+        Creates a ExtensionActor emulating an extension specified by the provided
+        extensionInfo and actionSet. Copies the dataFetcherBase to serve as the dynamic
+        loader of info required by a WireProtocolData* object.
+        """
+        self.action_scope = Mock()
+        ExtensionActor._configure_action_scope(self.action_scope, actionSet)
+
+        self.data_fetcher = dataFetcherBase.copy()
+        ExtensionActor._configure_data_fetcher(self.data_fetcher, actionSet, extensionInfo)
+
+        self.get_command = lambda cmd: getattr(self.action_scope, actionSet.getKeyForCommand(cmd), None)
```

Comment:
just curious: why is this defined as a lambda instead of a fuction?

Reply:
The ExtensionActor doesn't hold onto actionSet, so a closure within init is the only time it is in scope. Basically I made a decision not to hold on to actionSet (since we're computing action_scope), and getKeyFromCommand was the only thing required off of actionSet.